include/libtorrent/utp_socket_manager.hpp

```diff
@@ -130,23 +130,35 @@ namespace libtorrent
 		typedef std::multimap<std::uint16_t, utp_socket_impl*> socket_map_t;
 		socket_map_t m_utp_sockets;
 
+		using socket_vector_t = std::vector<utp_socket_impl*>;
+
 		// this is a list of sockets that needs to send an ack.
 		// once the UDP socket is drained, all of these will
 		// have a chance to do that. This is to avoid sending
 		// an ack for every single packet
-		std::vector<utp_socket_impl*> m_deferred_acks;
+		socket_vector_t m_deferred_acks;
+
+		// storage used for saving cpu time on "push_back"
+		// by using already pre-allocated vectors
+		socket_vector_t m_socket_drained_temp;
+
+		// prevents reentrant loop
+		bool m_in_socket_drained = false;
 
 		// sockets that have received or sent packets this
 		// round, may subscribe to the event of draining the
 		// UDP socket. At that point they may call the
 		// user callback function to indicate bytes have been
 		// sent or received.
-		std::vector<utp_socket_impl*> m_drained_event;
+		socket_vector_t m_drained_event;
 
 		// list of sockets that received EWOULDBLOCK from the
 		// underlying socket. They are notified when the socket
 		// becomes writable again
-		std::vector<utp_socket_impl*> m_stalled_sockets;
+		socket_vector_t m_stalled_sockets;
+
+		socket_vector_t m_writable_temp;
```

Comment:
I would expect it to be sufficient to just have a single temporary vector in here, that they all can use. The problem, as you pointed out, was that there were more than one libtorrent session going, which means the same static vector will get swapped from multiple threads and not be thread safe. By moving the temporary vector into here, you're back to thread-safety